File: dsa/9.stack_and_queue/20_Valid_Parentheses.py
import typing as t

class Solution:
    def isValid(self, s: str) -> bool:
        stack: t.List[int] = []
        bracket_dict: t.Dict[str, str] = {
            "}": "{"
            ,")": "("
            ,"]": "["
        }
        for char in s:
            if (
                char in bracket_dict.keys() 
                and stack
                and stack[-1] == bracket_dict.get(char)
            ):
                stack.pop()
            else:
                stack.append(char)
        
        if stack:
            return False
        else:
            return True

File: dsa/9.stack_and_queue/20_Valid_Parentheses.java
package stack_and_queue;
import java.util.*;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<Character>();
        Map<Character, Character> bracketMap = new HashMap<Character, Character>();
        bracketMap.put(')', '(');
        bracketMap.put(']', '[');
        bracketMap.put('}', '{');

        for (int i=0; i < s.length(); i++){
            if (stack.empty() == true 
                || bracketMap.containsKey(s.charAt(i)) == false 
                || stack.peek() != bracketMap.get(s.charAt(i))) {
                stack.push(s.charAt(i));
            }
            else {
                stack.pop();
            }
        }
        if (stack.empty() == true){
            return true;
        }
        return false;
    }
}